### ISSUE TITLE
`Conv` is 2x slow than pytorch `Conv` on cpu

### ISSUE THREAD

**Issue by yiyuezhuo** [FluxML/NNlib.jl#234]
As @ToucheSir suggested, the Discourse [post](https://discourse.julialang.org/t/conv-is-2x-slow-than-pytorch-conv-on-cpu/39802/8) is opened here as well. Also I have used `MKL.jl` as the suggestion.

Example:
```julia
using LinearAlgebra: BLAS
BLAS.vendor()  # :mkl

using BenchmarkTools
using Flux

conv = Conv((7,7), 3 => 64; stride=2, pad=3)
dummy = randn(Float32, 224, 224, 3, 2);
conv(dummy)  # ignore compile time
@benchmark conv(dummy) 

#=
BenchmarkTools.Trial: 
  memory estimate:  19.29 MiB
  allocs estimate:  45
  --------------
  minimum time:     5.384 ms (0.00% GC)
  median time:      5.825 ms (0.00% GC)
  mean time:        6.245 ms (5.38% GC)
  maximum time:     7.804 ms (17.60% GC)
  --------------
  samples:          800
  evals/sample:     1
=#

cor = Flux.CrossCor((7,7), 3 => 64; stride=2, pad=3)
dummy = randn(Float32, 224, 224, 3, 2);
cor(dummy)  # ignore compile time
@benchmark cor(dummy) 
#=
BenchmarkTools.Trial: 
  memory estimate:  19.29 MiB
  allocs estimate:  51
  --------------
  minimum time:     5.518 ms (0.00% GC)
  median time:      5.806 ms (0.00% GC)
  mean time:        6.042 ms (4.94% GC)
  maximum time:     8.333 ms (13.14% GC)
  --------------
  samples:          827
  evals/sample:     1
=#

@benchmark conv($dummy) 
#=
BenchmarkTools.Trial: 
  memory estimate:  19.29 MiB
  allocs estimate:  45
  --------------
  minimum time:     5.503 ms (0.00% GC)
  median time:      5.786 ms (0.00% GC)
  mean time:        6.026 ms (4.99% GC)
  maximum time:     8.327 ms (12.01% GC)
  --------------
  samples:          830
  evals/sample:     1
=#
```

It seems that `@benchmark` with `PyCall` will cause memory leak, so here're `@time`s with `PyCall` multiple times manually, but it's still clear that the 2x gap exists.

```julia
using PyCall
const torch = pyimport_conda("torch", "torch")
const nn = pyimport("torch.nn")
dummy_py = torch.randn(2, 3, 224, 224);
conv1_py = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=true)
conv1_py(dummy_py)

@time conv1_py(dummy_py) # 0.002447 seconds (3 allocations: 144 bytes)
@time conv1_py(dummy_py) # 0.002318 seconds (3 allocations: 144 bytes)
@time conv1_py(dummy_py) # 0.002344 seconds (3 allocations: 144 bytes)
@time conv1_py(dummy_py) # 0.002278 seconds (3 allocations: 144 bytes)
@time conv1_py(dummy_py) # 0.002603 seconds (3 allocations: 144 bytes)

@time conv(dummy) # 0.015352 seconds (45 allocations: 19.287 MiB, 49.63% gc time)
@time conv(dummy) # 0.005321 seconds (45 allocations: 19.287 MiB)
@time conv(dummy) # 0.005424 seconds (45 allocations: 19.287 MiB)
@time conv(dummy) # 0.005461 seconds (45 allocations: 19.287 MiB)
@time conv(dummy) # 0.005627 seconds (45 allocations: 19.287 MiB)
```

Though memory usage of PyCall looks fishy, @ToucheSir's [result](https://discourse.julialang.org/t/conv-is-2x-slow-than-pytorch-conv-on-cpu/39802/8?u=yiyuezhuo) in Python gives a similar result.

**Comment by ToucheSir** [comment]
To add on my own findings, relative performance is unchanged if one uses `torch.nn.functional.conv2d` and `NNlib.conv` instead of the higher-level layers.

> memory usage of PyCall looks fishy

IIRC BenchmarkTools can't track any allocations made by Python, so the 144 bytes is probably just Julia-side bookkeeping.

**Comment by CarloLucibello** [comment]
I looked into this, see perf results in #695 #696. The ~2x gap is **mostly a threading artifact**, not the im2col+GEMM algorithm itself.

**Root cause:** NNlib's CPU conv parallelizes *only over the batch dimension* ([`conv_im2col.jl`](https://github.com/FluxML/NNlib.jl/blob/master/src/impl/conv_im2col.jl) partitions `axes(x, 5)` and `@spawn`s one task per batch chunk). PyTorch dispatches float32 convs to **oneDNN**, which parallelizes *within* a single image. So with the reported `batch=2` and ≥4 threads, NNlib keeps 2 cores busy while oneDNN uses all of them.

**Evidence** — per-image time for the issue's 7x7 case (4 threads):

| batch | 1 | 2 | 4 | 8 |
|---|---|---|---|---|
| NNlib (ms/img) | 6.46 | 4.76 | 2.76 | **2.39** |
| PyTorch (ms/img) | 2.58 | 2.46 | 2.51 | **2.41** |

NNlib only improves as the batch grows; PyTorch is flat. **Once `batch ≥ nthreads` they match** — and at `batch=8` NNlib is actually slightly *faster* than oneDNN on this 7x7 first-layer conv. So the 2x is purely idle cores at small batch.

A **residual ~1.6x gap remains for 3x3 convs even when saturated**, which *is* algorithmic: a 3x3 conv expands the input 9x (`k²`) into the col buffer, and that memory traffic is something oneDNN's direct conv never pays.

**Suggested fix:** parallelize *within* an image — partition the output spatial dimension (the GEMM's `M`) across threads in `conv_im2col.jl`, instead of only over the batch axis. This closes the small-batch/batch=1 gap entirely. Tiling/fusing im2col+GEMM (per-thread tile instead of one big buffer) would additionally narrow the 3x3 case; a full direct-conv rewrite isn't worth it.

Reproducible benchmarks + full analysis are in [`benchmark/conv_vs_pytorch/`](https://github.com/FluxML/NNlib.jl/tree/master/benchmark/conv_vs_pytorch).
